### PR TITLE
fix: extract JUnit failure body text when message attribute is absent

### LINKED ISSUE #216: fix: Empty error field when JUnit failure text is in element body instead of message attribute
## Problem

When a JUnit XML test failure has the error message inside the `<failure>` element body (not in a `message` attribute), the `error` field in the analysis result is empty.

### JUnit XML structure:
```xml
<testcase name="...[test_id:1514] should fail pinging...">
    <failure type="Failure">tests/network/networkpolicy.go:136
Timed out after 15.001s.
Expected failure, but got no error.
tests/network/networkpolicy.go:137</failure>
</testcase>
```

Note: there is no `message="..."` attribute on the `<failure>` tag — the text is the element body.

### Expected
The `error` field should contain: `Timed out after 15.001s. Expected failure, but got no error.`

### Actual
The `error` field is empty (`""`). The text only appears in `artifacts_evidence`.

### Likely cause
The JUnit parser only reads the `<failure message="...">` attribute but does not fall back to reading the `<failure>` element body text when the attribute is absent.

## Done

- [ ] JUnit parser extracts `<failure>` element body when `message` attribute is absent
- [ ] Error field populated correctly for body-only failures
- [ ] Backend tests pass

### PR DESCRIPTION
**Pull request by myakove**
## Summary

Fixes #216 — Fall back to reading the `<failure>` element body text when the `message` attribute is absent, instead of returning an empty string.

## Problem / Motivation

Some JUnit XML reports emit failure details only as body text within the `<failure>` element, without a `message` attribute. Previously, the parser returned an empty string in these cases, causing failure information to be silently lost during analysis.

## Changes

- When the `message` attribute is missing from a `<failure>` element, the parser now reads the element's body text as the failure message.
- Ensures failure details are always captured regardless of how the JUnit XML is structured.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved extraction of error messages/details: ignores leading blank lines/whitespace, prefers explicit message fields, falls back to summarizing stack traces while skipping file:line noise, and safely handles missing/empty fields.

* **Tests**
  * Expanded coverage for failure parsing edge cases: leading whitespace/newlines, message-vs-body precedence, stack-trace fallback, exclusion of passed tests, and nested report structures.

### CODE REVIEW

File: src/jenkins_job_insight/xml_enrichment.py [hunk 1]
```diff
@@ -52,7 +64,7 @@ def extract_failures_from_xml(raw_xml: str) -> list[dict[str, str]]:
             {
                 "test_name": test_name,
                 "error_message": result_elem.get("message", "")
-                or ((result_elem.text or "").split("\n")[0].strip()),
+                or _first_nonempty_line(result_elem.text or ""),
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/xml_enrichment.py]: Fixed — added .strip() so whitespace-only message attributes fall through to the body text fallback.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

File: src/jenkins_job_insight/analyzer.py [hunk 2]
```diff
@@ -867,11 +867,24 @@ def extract_failures_from_test_report(test_report: dict) -> list[TestFailure]:
...
+                            error_details = stripped
+                            break
+
                 failures.append(
                     TestFailure(
                         test_name=full_name,
-                        error_message=case.get("errorDetails", "") or "",
-                        stack_trace=case.get("errorStackTrace", "") or "",
+                        error_message=error_details,
+                        stack_trace=stack_trace,
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/analyzer.py]: Fixed — added .strip() to errorDetails and broadened file:line regex from .go to .\w+ to cover Java, Python, etc.

File: tests/test_analyzer.py [hunk 3]
```diff
@@ -3049,3 +3050,173 @@ def test_recover_without_code_fields(self) -> None:
...
+                    "name": "t",
+                    "status": "FAILED",
+                    "errorDetails": "",
+                    "errorStackTrace": "existing trace with details",
+                }
+            ]
+        )
+        failures = extract_failures_from_test_report(report)
+        assert failures[0].error_message == "existing trace with details"
+        assert failures[0].stack_trace == "existing trace with details"
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/test_analyzer.py:3163]: Fixed — added test_whitespace_only_error_details_falls_back_to_stack_trace regression test.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `src/jenkins_job_insight/analyzer.py:873` (nitpick) — Extract this fallback parser into a shared helper to keep error normalization consistent.
> Addressed: Fixed — extracted _derive_error_details() helper. Kept separate from _first_nonempty_line in xml_enrichment.py due to different semantics.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review
